### PR TITLE
Implement VPAWModel, to run segmentation/geometry pipeline on new cases

### LINKED ISSUE #11: make VPAWModel run segmentation/geometry pipeline on new cases
Give VPAWModel the ability to run the [pediatric_airway_atlas](https://github.com/uncbiag/pediatric_airway_atlas) pipeline on new cases

### PR DESCRIPTION
**Pull request by Leengit**
Closes #11.  The VPAWModel button in the application is now connected to a user interface and implementation that supports:

1. Install `pediatric_airway_atlas` and all of its Python package dependencies.  Because the repository for it is not public, the user is able to access the top-level code after specifying its location on the local computer.  The Python package dependencies are installed into the environment maintained by the underlying 3D Slicer apparatus and to not affect global or virtual Python environments.
2. Run the full pipeline.  Inputs to be specified are (1) the input / output directory (called `vpaw-root-dir` in some of our examples), which contains `images/*`, `landmarks/*`, and `FilteredControlBlindingLogUniqueScanFiltered.xls`, and (2) the models directory (called `models` in some of our examples), which contains a file with a name like `116(158.10-38.AM.24.Mar).pth`.

### CODE REVIEW

File: Modules/Scripted/VPAWModel/VPAWModel.py [hunk 1]
```diff
@@ -377,58 +414,287 @@ def __init__(self):
...
+        try:
+            # Change directory and create files
+            os.chdir(self.pediatric_airway_atlas_directory)
+            ConfigHandle, ConfigName = tempfile.mkstemp(suffix=".yaml", text=True)
+            SegmentHandle, SegmentName = tempfile.mkstemp(suffix=".yaml", text=True)
+
+            # Add text to the main configuration file
+            b_s_f_s = "False"
+            n_p = 1
+            os.write(
```
Comment (ebrahimebrahim) [inline Modules/Scripted/VPAWModel/VPAWModel.py:610]: It would be better to read and write yaml using pyyaml.

This would make it easier to extend configurability later, which I imagine is something we will want to do.

File: Modules/Scripted/VPAWModel/VPAWModel.py [hunk 2]
```diff
@@ -377,58 +414,287 @@ def __init__(self):
...
+use_planes_for_laplace_marking: True
+balance_spacing_for_segmentation: {b_s_f_s}
+num_processes: {n_p}
+""".encode(),
+            )
+            os.close(ConfigHandle)
+            ConfigHandle = None
+
+            # Add text to the configuration file for segmentation
+            os.write(
```
Comment (ebrahimebrahim) [inline Modules/Scripted/VPAWModel/VPAWModel.py:631]: Same here, I would say read and write yaml using pyyaml

File: Modules/Scripted/VPAWModel/VPAWModel.py [hunk 3]
```diff
@@ -253,31 +273,50 @@ def updateGUIFromParameterNode(self, caller=None, event=None):
...
+        self.ui.VPAWModelsDirectory.toolTip = (
+            "Directory containing file named like '116(158.10-38.AM.24.Mar).pth'"
+        )
+        if os.path.isdir(self.ui.VPAWRootDirectory.currentPath) and os.path.isdir(
+            self.ui.VPAWModelsDirectory.currentPath
+        ):
+            self.ui.runPediatricAirwayAtlasButton.toolTip = (
+                "Run Pediatric Airway Atlas (This may take many hours)"
+            )
+            self.ui.runPediatricAirwayAtlasButton.enabled = True
```
Comment (ebrahimebrahim) [inline Modules/Scripted/VPAWModel/VPAWModel.py:313]: The enabled state of this button seems to toggle on properly, but it does not toggle off properly.

If I fill in directories for both `VPAWRootDirectory` and `VPAWModelsDirectory` then the button is enabled as expected. However if I then delete one of them and set it to be blank, the button remains enabled.

I am not sure why this is happening, maybe something to do with the triggers for `updateGUIFromParameterNode` to get called... not sure what triggers it exactly

File: Modules/Scripted/VPAWModel/Resources/UI/VPAWModel.ui [hunk 4]
```diff
@@ -38,181 +38,94 @@
...
-         <double>-100.000000000000000</double>
-        </property>
-        <property name="maximum">
-         <double>500.000000000000000</double>
+         <string>Install Pediatric Airway Atlas and Dependencies</string>
         </property>
-        <property name="value">
-         <double>0.500000000000000</double>
+        <property name="text">
+         <string>Install Pediatric Airway Atlas and Dependencies</string>
```
Comment (ebrahimebrahim) [inline Modules/Scripted/VPAWModel/Resources/UI/VPAWModel.ui:72]: The name of this button is a bit misleading. It does two things:

- installs dependencies into the python environment
- adds paa to the `sys.path`

The first one is "installing" and we only need to do it once.

The second one is not really "installing" and we need to do it every time.

The name "Install" of this button makes it sound like we only need to do it once.

We should come up with a clearer name. One suggestion: "Link to Pediatric Airway Atlas Codebase"

**Review by ebrahimebrahim — CHANGES_REQUESTED** [review]
My tests were successful! I ran this on a root directory with a couple of images and landmarks. It ran the pipeline with no trouble. Then I switch to VPAW visualize and saw everything is there as expected.

A couple of things that we should definitely do before we call this done:

- We need to be able to run on individual cases. That's more likely how this would be used. So there should be a way to choose a specific patient ID to run the inference on. This may require some change in `pediatric_airway_atlas` but it shouldn't be too too bad.
- For the user I think it is *much* more convenient if VPAWModel remembers the previously used values for `VPAWRootDirectory` and `PediatricAirwayAtlasDirectory` and `VPAWModelsDirectory`. (i.e. when restarting the application the previously used values should be remembered.) I am not sure how to do this but I think for VPAWModel it is worth figuring out. We can port the same solution over to VPAWVisualize to make it remember the last used root directory. If it's too hard we can forget it, but I bet there is a way to do in Slicer.

File: Modules/Scripted/VPAWModel/VPAWModel.py [hunk 5]
```diff
@@ -377,58 +414,287 @@ def __init__(self):
...
+            )
+            if not want_install:
+                mesg = f"Package{plural} installation declined; giving up."
+                slicer.util.errorDisplay(mesg, "Install Error")
+                print(mesg)
+                return False
+            try:
+                # Install missing packages
+                with BusyCursor():
+                    slicer.util.pip_install(["-U", "pip", "setuptools", "wheel"])
```
[the remark this answers is not in the document]
Comment (Leengit) [reply Modules/Scripted/VPAWModel/VPAWModel.py:532]: That's `-U` for `--upgrade`.  Upgrading these gets the latest installation software, for installing the dependencies of `pediatric_airway_atlas`.

File: Modules/Scripted/VPAWModel/VPAWModel.py (hunk 3 above)
Comment (Leengit) [reply Modules/Scripted/VPAWModel/VPAWModel.py:313]: Good catch.  I have a fix: changing the string in one of those widgets to blank doesn't signal `currentPathChanged`, but it does signal `validInputChanged`.  So, now we'll connect both signals to the button enabling/disabling code.

File: Modules/Scripted/VPAWModel/Resources/UI/VPAWModel.ui (hunk 4 above)
Comment (Leengit) [reply Modules/Scripted/VPAWModel/Resources/UI/VPAWModel.ui:72]: Will do.

File: Modules/Scripted/VPAWModel/VPAWModel.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply Modules/Scripted/VPAWModel/VPAWModel.py:532]: Oh my bad you are right. I saw it uninstalling in order to do the upgrade and jumped to conclusions

File: Modules/Scripted/VPAWModel/VPAWModel.py (hunk 1 above)
Comment (Leengit) [reply Modules/Scripted/VPAWModel/VPAWModel.py:610]: Good thought.  Will do.

File: Modules/Scripted/VPAWModel/VPAWModel.py (hunk 2 above)
Comment (Leengit) [reply Modules/Scripted/VPAWModel/VPAWModel.py:631]: Will do.

**Comment by Leengit** [comment]
The commit I just pushed, 266efe2415fc4ffd4c9ee5a6d7345f8b59b433e4, supports filtering the set of subjects by the prefix of the file name.  However, it requires a modification to `pediatric_airway_atlas` to have this functionality; at present this modification is available as branch `subject_prefix` but is not merged into `master`.  Without the new functionality of `pediatric_airway_atlas`, the `VPAWModel` module will give an error if a non-blank "Patient prefix" is provided.

**Comment by ebrahimebrahim** [comment]
That's a great solution and I think solves the problem completely. In our readme instead of instructing the user to
```sh
git clone git@github.com:uncbiag/pediatric_airway_atlas.git
```
we can just instruct them to clone and check out the correct branch right away
```sh
git clone -b subject_prefix git@github.com:uncbiag/pediatric_airway_atlas.git
```

if pediatric_airway_atlas later wants to merge that to main then they can do that.

**Comment by Leengit** [comment]
Regarding preserving the values in the form widgets in VPAWModel and VPAWVisualize: what we have now is the first alternative below.  What do we want?
1. If one saves the scene (File->Save Data, or a shortcut) then a restart of VPAW and loading the scene (File->Add Data, or a shortcut) restores the values to the form widgets in the various modules.  (I suppose we could modify this by calling Save Data every time there is a change, without explicit user consent.  Slow?  Side effects?)
2. Instead of saving to the scene, save to QSettings, which are application-wide rather than scene-specific.  (**Plus**: this would not require explicit scene saving and loading; we'd `QSettings.setValue(self, key, value)` each time the user changes a value and we'd get the `QSettings.value(self, key)` as part of normal application start up.  Alternatively, we'd save out the current values only when the user clicks on a button indicating "Save as default".  **Minus**: the user couldn't have separate values for separate scenes.)
3. Somehow have both scene-specific and application-wide values for these widgets.  Maybe: if both scene-specific and application-wide sources have values then we'd let the scene-specific values dominate.  Any update to the widgets would be saved as both scene-specific (at the time of scene saving) and as application-wide (immediately; or when "Save as Default").  (**Plus**: it's flexible.  **Minus**: it's confusing.)

Or some mixture??: such as `pediatric_airway_atlas` and `models` would be application-wide but `vpaw-root-dir` and patient prefix would be scene-specific.

**Comment by ebrahimebrahim** [comment]
I think (2) is simple, convenient, and makes the most sense. The root directory, the repository location, and the model directory are all really "application settings" and would likely not change from scene to scene.

Regarding (1): Saving and loading mrml scenes is likely not the primary way in which VPAW will be used. I can imagine that being used to save particular views, but I don't imagine that the users will typically want to load a mrml scene when using VPAWModel.

Regarding (3): I agree that it's a bit confusing, and perhaps overengineering for the current application.

A mixture would also be confusing and I feel that the problem is fully solved by the `QSettings` approach.

File: Modules/Scripted/VPAWModel/VPAWModel.py [hunk 6]
```diff
@@ -361,74 +381,395 @@ def onApplyButton(self):
...
+            This is the input/output directory, e.g., "path/to/vpaw-data-root".  It must
+            contain FilteredControlBlindingLogUniqueScanFiltered.xls, images/*, and
+            landmarks/*.
+        vPAWModelsDirectory : str
+            It must contain a file with name like "116(158.10-38.AM.24.Mar).pth".
+        patientPrefix : str
+            Process only files with this prefix.  Blank means all files.
+        """
+
+        # If self.pediatric_airway_atlas is not yet set, then see if we can set it.
```
Comment (ebrahimebrahim) [inline Modules/Scripted/VPAWModel/VPAWModel.py:571]: I like that it now tries to link if I forget to press "link" before pressing "run"

File: Modules/Scripted/VPAWModel/VPAWModel.py [hunk 7]
```diff
@@ -361,74 +381,395 @@ def onApplyButton(self):
...
+                        [
+                            f"--images_dir={images_dir}",
+                            f"--input_landmarks_dir={input_landmarks_dir}",
+                            f"--output_landmarks_dir={output_landmarks_dir}",
+                            f"--num_workers={num_workers}",
+                            f"--subject_prefix={subject_prefix}",
+                        ],
+                    )
+                except:
+                    slicer.util.errorDisplay(
```
Comment (ebrahimebrahim) [inline Modules/Scripted/VPAWModel/VPAWModel.py:623]: This error branch seems to capture the case where I put in a bad subject ID, i.e one for which there is no input image. The error message is then confusing in that case

The message could be updated to say "It could also be that you entered a subject prefix that does not exist" in order to cover that possibility

Or there can be a separate error handling for a bad subject ID

**Review by ebrahimebrahim — APPROVED** [review]
Really really wonderful work on this. All points are addressed, the code looks good, and it builds and works perfectly for me.

Added just one minor comment; once you are happy with that please feel free to merge!

File: Modules/Scripted/VPAWModel/VPAWModel.py (hunk 6 above)
Comment (Leengit) [reply Modules/Scripted/VPAWModel/VPAWModel.py:571]: Yes, with the `pediatric_airway_atlas` value automatically repopulating at application start, I wanted a mechanism whereby the `sys.path` linking to it was also automated.  But in the case where it was populated but dependencies hadn't yet been installed -- if the user had typed in a value but hadn't hit the Link button -- I didn't want to force the user through the dependency installation at application restart.  Waiting until the user hits a button (`Run` or `Link`) was a nice way to achieve these priorities.

File: Modules/Scripted/VPAWModel/VPAWModel.py (hunk 7 above)
Comment (Leengit) [reply Modules/Scripted/VPAWModel/VPAWModel.py:623]: So far, I have been able to maintain a clean separation between `VPAW` and `pediatric_airway_atlas` in that only the latter ever cares about the subdirectory organization of and file contents within the `pediatric_airway_atlas/`, `models/`, or `vpaw-root-dir/` directories.  (Well, `VPAW` does know the names and locations of the top-level `pediatric_airway_atlas` Python modules that it is calling.) Rather than have `VPAW` look inside, checking for the existence of a particular patient prefix by knowing where to look in `vpaw-root-dir/`, I'll simply modify the error message along the lines that you suggest.

Comment (Leengit) [reply Modules/Scripted/VPAWModel/VPAWModel.py:623]: Actually, in my testing, a bad patient prefix does not trigger any error.  The subject_prefix-enabled pipeline happily processes all zero files that it finds with that prefix.  That is, it happily does nothing.  So, I'm not so sure that the modified error messages (just pushed) are a net win.

Comment (ebrahimebrahim) [reply Modules/Scripted/VPAWModel/VPAWModel.py:623]: Oh, hmm. In my test it did error out when I did that. Let me try again...

Comment (ebrahimebrahim) [reply Modules/Scripted/VPAWModel/VPAWModel.py:623]: Hmm, okay so if the pipeline has already processed all images then when I put a bogus subject ID I get the same result as you: it happily processes zero files. But if the pipeline has some unprocessed images in there, then a bogus subject ID results in the error message.

Comment (Leengit) [reply Modules/Scripted/VPAWModel/VPAWModel.py:623]: Good sleuthing!  I'll go ahead and merge with these messages modified as you suggest.
